plugins/cfu/fu-cfu-module.c

```diff
@@ -0,0 +1,110 @@
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-cfu-module.h"
+
+struct _FuCfuModule {
+	FuDevice parent_instance;
+	guint8 component_id;
+	guint8 bank;
+};
+
+G_DEFINE_TYPE(FuCfuModule, fu_cfu_module, FU_TYPE_DEVICE)
+
+static void
+fu_cfu_module_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuCfuModule *self = FU_CFU_MODULE(device);
+	fu_common_string_append_kx(str, idt, "Bank", self->bank);
+	fu_common_string_append_kx(str, idt, "ComponentId", self->component_id);
+}
+
+gboolean
+fu_cfu_module_setup(FuCfuModule *self, const guint8 *buf, gsize bufsz, gsize offset, GError **error)
+{
+	FuDevice *parent = fu_device_get_proxy(FU_DEVICE(self));
+	guint32 version_raw = 0;
+	guint8 tmp = 0;
+	g_autofree gchar *instance_id0 = NULL;
+	g_autofree gchar *instance_id1 = NULL;
+	g_autofree gchar *instance_id2 = NULL;
+	g_autofree gchar *logical_id = NULL;
+	g_autofree gchar *name = NULL;
+	g_autofree gchar *version = NULL;
+
+	/* component ID */
+	if (!fu_common_read_uint8_safe(buf, bufsz, offset + 0x5, &self->component_id, error))
+		return FALSE;
+	name = g_strdup_printf("%s (0x%02X)", fu_device_get_name(parent), self->component_id);
+	fu_device_set_name(FU_DEVICE(self), name);
+
+	/* these GUIDs may cause the name or version-format to be overwritten */
+	instance_id0 = g_strdup_printf("USB\\VID_%04X&PID_%04X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)));
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id0);
+	instance_id1 = g_strdup_printf("USB\\VID_%04X&PID_%04X&CID_%02X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)),
+				       self->component_id);
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id1);
+
+	/* bank */
+	if (!fu_common_read_uint8_safe(buf, bufsz, offset + 0x4, &tmp, error))
+		return FALSE;
+	self->bank = tmp & 0b11;
+	instance_id2 = g_strdup_printf("USB\\VID_%04X&PID_%04X&CID_%02X&BANK_%01X",
+				       fu_udev_device_get_vendor(FU_UDEV_DEVICE(parent)),
+				       fu_udev_device_get_model(FU_UDEV_DEVICE(parent)),
+				       self->component_id,
+				       self->bank);
+	fu_device_add_instance_id(FU_DEVICE(self), instance_id2);
+
+	/* version */
+	if (!fu_common_read_uint32_safe(buf, bufsz, offset, &version_raw, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	fu_device_set_version_raw(FU_DEVICE(self), version_raw);
+	version = fu_common_version_from_uint32(version_raw,
+						fu_device_get_version_format(FU_DEVICE(self)));
+
+	/* logical ID */
+	logical_id = g_strdup_printf("CID:0x%02x", self->component_id);
+	fu_device_set_logical_id(FU_DEVICE(self), logical_id);
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_cfu_module_init(FuCfuModule *self)
+{
+	fu_device_add_protocol(FU_DEVICE(self), "com.microsoft.cfu");
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_QUAD);
```

Comment:
From spec this is wrong:
* This specification does not mandate any specific format for the firmware version. See section Firmware Version for guidelines.

Reply:
Agreed, but I wanted to at least choose a sane default. Given we need a VID/PID entry in the quirk I figured we could override the `VersionFormat` there, and that's why I added the instance IDs before reading and processing the version `ui32_t`.